cpp/doc/libcuspatial_refactoring_guide.md

```diff
@@ -0,0 +1,275 @@
+     type of data must be done at run time.
+  2. All inputs are arrays of scalars. Longitude and latitude are separate. 
+  3. The output is a returned `unique_ptr<cudf::column>`.
+  4. The output is allocated inside the function using the passed memory resource.
+  5. The public API does not take a stream. There is a `detail` version of the API that takes a
+     stream. This follows libcudf, and may change in the future.
+
+## File Structure
+
+For now, libcuspatial APIs should be defined in a header file in the
+`cpp/include/cuspatial/experimental/` directory. Later, as we adopt the new API, we will rename
+the `experimental` directory. The API header should be named after the API. In the example, 
+`haversine.hpp` defines the `cuspatial::haversine_distance` API.
+
+The implementation must also be in a header, but should be in the `cuspatial/experimental/detail`
+directory.  The implementation should be included from the API definition file, at the end of the 
+file. Example:
+
+```C++
+... // declaration of API above this point
+#include <cuspatial/experimental/detail/haversine.hpp>
+```
+
+## Namespaces
+
+Public APIs are in the `cuspatial` namespace. Note that both the header-only API and the libcudf-
+based API can live in the same namespace, because they are non-ambiguous (very different 
+parameters).
+
+Implementation of the header-only API should be in a `cuspatial::detail` namespace.
+
+## Implementation
+
+The main implementation should be in detail headers. 
+
+### Header-only API Implementation
+
+Because it is a statically typed API, the header-only implementation can be much simpler than the 
+libcudf-based API, which requires run-time type dispatching. In the case of `haversine_distance`, it is
+a simple matter of a few static asserts and dynamic expectation checks, followed by a call to
+`thrust::transform` with a custom transform functor.
+
+```C++
+template <class LonLatItA,
+          class LonLatItB,
+          class OutputIt,
+          class Location = typename std::iterator_traits<LonLatItA>::value_type,
+          class T        = typename Location::value_type>
+OutputIt haversine_distance(LonLatItA a_lonlat_first,
+                            LonLatItA a_lonlat_last,
+                            LonLatItB b_lonlat_first,
+                            OutputIt distance_first,
+                            T const radius,
+                            rmm::cuda_stream_view stream)
+{
+  using LocationB = typename std::iterator_traits<LonLatItB>::value_type;
+  static_assert(std::conjunction_v<std::is_same<location_2d<T>, Location>,
+                                   std::is_same<location_2d<T>, LocationB>>,
+                "Inputs must be cuspatial::location_2d");
+  static_assert(
+    std::conjunction_v<std::is_floating_point<T>,
+                       std::is_floating_point<typename LocationB::value_type>,
+                       std::is_floating_point<typename std::iterator_traits<OutputIt>::value_type>>,
```

Comment:
Should we also assert that the precision of these types align?

Reply:
Do you mean that the `T` in each of the `Location` types is the same (`float`+`float`, `double`+`double` etc.?)  The `static_assert` just before this one ensures this is the case. 

```
 static_assert(
    std::conjunction_v<std::is_same<lonlat_2d<T>, Location>, std::is_same<lonlat_2d<T>, LocationB>>,
    "Inputs must be cuspatial::lonlat_2d");
```

If either is no `lonlat_2d<T>` then it will fail.